### PR TITLE
CDA Loader

### PR DESCRIPTION
**Pull request by jbkolze**
In its current form, the CDA loader requires two options: a criteria file linking SHEF values to CWMS time series IDs as well as a CDA api-key.  The criteria file is parsed and saved as a set of ShefTransform objects.  The api-key is appended to all POST requests for authorization purposes.

In load_time_series(), SHEF values are initially parsed and grouped into POST payloads based on time series ID, owning office, and units.  Conceptually, the SHEF values should probably be loaded in this method, but due to the nature of some SHEF products (see: [LPMS product](https://github.com/HydrologicEngineeringCenter/SHEF_processing/blob/feature/cda_loader/example-shef-files/lrl/cwbi_lpms.shef.2024.06.25.080120)) this can result in a *lot* of unnecessary POST requests to CDA.  Therefore, payloads are grouped to cut down on requests as much as possible.

All of the actual "output" actually occurs in the done() method, which isn't ideal.  There currently isn't an obvious way to create a workflow like "ingest all SHEF data, group as necessary, then load".  For these purposes it might be beneficial to have an optional base_loader method that runs between load_time_series() and done().  Running the output in done() doesn't cause issues with the current setup, but that won't remain the case if unloading functionality is added to the CDA Loader.

During the final POST task creation step (parse_payload_tasks()), payloads are split up into continuous series of timestamps based on the time series interval in order to avoid deleting existing data due to gaps in the POST request.  Further investigation may be useful to determine if this is actually necessary depending on the CDA store_rule utilized for the POST request.

The CWMSpy library is used for all CDA requests.  Because this is an extremely slow process if handled synchronously, all CWMSpy requests are wrapped in asyncio.to_thread() calls that are eventually unwrapped in an asyncio.as_completed() loop.

Loader can be updated to pull ShefTransform data directly from CWMS once the structure has been identified and set up.

And a quick side-note... It might be helpful to enforce some form of auto-formatting in the project, e.g. black, ruff, etc.  This loader is contained to one file so no issues currently, but might become a problem when multiple users are tweaking the same files.

### CODE REVIEW

**Comment by jbkolze** [comment]
> What version of Python do we want to require? My Rocky Linux VM installs 3,9 by default but my Ubuntu VM installs 3.10. Of course either can be updated manually to later versions. The App Portal has 3.11 and 3.12 for Windows.
> 
> The reason I'm asking is mypy on Rocky Linux flagged two type hints as requiring version 3.10 in cda_loader (lines 83 and 259, using `var: type1 | type2`). In 3.9 this has to be `var: Union[type1, type2]`.

I believe cwms_python is targeting Python 3.9, so this should probably be changed to a Union.  Should've thought to downgrade my interpreter to avoid this -- my bad.  Thanks for catching.  I will get it fixed.

**Comment by perrymanmd** [comment]
> > What version of Python do we want to require? My Rocky Linux VM installs 3,9 by default but my Ubuntu VM installs 3.10. Of course either can be updated manually to later versions. The App Portal has 3.11 and 3.12 for Windows.
> > The reason I'm asking is mypy on Rocky Linux flagged two type hints as requiring version 3.10 in cda_loader (lines 83 and 259, using `var: type1 | type2`). In 3.9 this has to be `var: Union[type1, type2]`.
> 
> I believe cwms_python is targeting Python 3.9, so this should probably be changed to a Union. Should've thought to downgrade my interpreter to avoid this -- my bad. Thanks for catching. I will get it fixed.

I would've just commented to change those lines in a review if I was sure that 3.9 was the correct target, but I wasn't sure if we should target something later.

**Comment by Enovotny** [comment]
yes 3.9 is the target.  Right now CDA is not working well with 3.10 and above.  I can't figure out how to get the certificates to work.  Keep getting SSL errors.  Works fine in 3.9.  Although the new airflow will be running 3.11 so we will need to figure that out.

**Comment by perrymanmd** [comment]
As for handling badly-ordered SHEF message (causing too many calls to `load_timeseries()`), The loader could take an argument for the number of time series to collect simultaneously (currently it is just one). Say the number was 10 - `base_loader` would collect up to 10 time series simultaneously and when an 11th time series was encountered, `load_timeseries()` would be called with the oldest of the current time series before starting the collection of the new time series. We would also probably want a specifiable limit of the number of values to collect in any time series before calling `load_timeseries()` regardless of whether the limit on simultaneous time series had been reached (e.g., if any time series reaches 100 or 1000 values, load it regardless).

**Comment by Enovotny** [comment]
In the cloud the way I see this working is the following.  Airflow grabs a file.  Parser parses it.  Either puts data into dataframe/list/file or whatever format we want to use, we can add this part in airflow. Loader accepts the object/dataframe. If it is a dataframe can sort on time series ID merge with information in crit file/timeseries group. and then send data to database by time series id using cwms-python.  Each run in airflow would be to process a single file.  So by the time the loader gets the data we already have all of the data through the parsers and already know what time series are going to be saved and how many values were in the file.

**Comment by perrymanmd** [comment]
> In the cloud the way I see this working is the following. Airflow grabs a file. Parser parses it. Either puts data into dataframe/list/file or whatever format we want to use, we can add this part in airflow. Loader accepts the object/dataframe. If it is a dataframe can sort on time series ID merge with information in crit file/timeseries group. and then send data to database by time series id using cwms-python. Each run in airflow would be to process a single file. So by the time the loader gets the data we already have all of the data through the parsers and already know what time series are going to be saved and how many values were in the file.

Yeah - I forgot about separating the loader from the parser. Right now the loader doesn't have `main()` but is called from the parser. Is there an actual reason to separate them? I know I originally had that model in mind but currently the loader is invoked using command line parameters to `shefParser` and the entire thing runs as a single script.

If I remember correctly, we'll also have to find a way to convey parameters other than the command line. If that's true then we'll need an environment to test it out in.

**Comment by jbkolze** [comment]
> As for handling badly-ordered SHEF message (causing too many calls to `load_timeseries()`), The loader could take an argument for the number of time series to collect simultaneously (currently it is just one). Say the number was 10 - `base_loader` would collect up to 10 time series simultaneously and when an 11th time series was encountered, `load_timeseries()` would be called with the oldest of the current time series before starting the collection of the new time series. We would also probably want a specifiable limit of the number of values to collect in any time series before calling `load_timeseries()` regardless of whether the limit on simultaneous time series had been reached (e.g., if any time series reaches 100 or 1000 values, load it regardless).

This sounds like a good solution to me.  I was envisioning a workflow such as `load_timeseries()` for the whole file and then `process_timeseries()` when applicable before `done()`, but that's a bit of an all-or-nothing solution.  Your proposal would allow processing to be tailored to each SHEF product and/or loader where needed.

**Comment by jbkolze** [comment]
Regarding running in one shefParser call vs. generating a middle-man product, I think either route will get us where we need to be.  Similar to Mike, though, I'm curious if there's a specific benefit you are seeing to splitting the parsing and the loading, Eric?  I feel like avoiding additional intermediate steps might cut down on maintenance issues in the long run, but I might be missing some context for the Airflow environment specifically.

**Comment by Enovotny** [comment]
In my mind I guess I would see things this way.  the shef parser could be it's own module.  Even placed in pypi as other's might be interested in it.  it outputs a dataframe or .out file. Then the loader could be simplified. If you are interacting with a dataframe I think a number of the functions you have in your code could be cut out or reduced to a single line.  filter duplicates,  grouping values together, datetime adjustments etc.  you would then have the entire dataset and could simply grab the subset you need for a given time series merge it with the information in the crit file/time series group, use the cwms-python function timeseries_df_to_json and then write it to the database.  This would eliminate the need to track the number of values per time series, would allow for a single CDA call per time series per file, and would separate tasks parse/load allowing us to track bugs easier. also keeps to parser code out of airflow and in an individual module.  What I see in airflow would be

df = parser_funct(abcd.SHEF, ...........)

loader_func(df, .......)

I don't know if things can be split out like that without a big reorganization and impacting the dss loader. but for airflow at the minimum we will need to have function calls and not calls from the command line.

**Comment by jbkolze** [comment]
I can see the value in splitting the parser and the loaders since that will move us more toward a micro-services approach rather than one monolithic application, which should theoretically be beneficial in the cloud.  I do think there's some coupling between the loaders and the parser that would have to be figured out, but Mike can obviously speak better than I can as to how big of a hassle that would be.

I think if we go this route it would make sense to decide on a "default" product, i.e. parsers all convert from something to *default* and loaders all load from *default* to something else.  I'm a little leery of using a dataframe as the "default" product (or rolling in pandas as a dependency in general) just because it's a pretty massive package to require for the relatively small amount of processing we're generally doing, especially for things like this that could hypothetically live in something like a Lambda function.

At the end of the day I think either approach is valid, but I do think whatever we decide to do here will likely set the direction of other parsers and loaders to follow, so it's probably worth putting some thought into.  On the other hand, might just be a case where we should throw something at the proverbial Airflow wall and see what sticks.  Don't know about you all, but I don't have enough cloud data engineering experience to anticipate exactly what hurdles we're going to run into in our new environment.

**Comment by perrymanmd** [comment]
> pandas is already rolled into cwms-python so it will be needed anyways and is already rolled into airflow. I can see the parser not using pandas and just saving it as a csv/text file and then in the loader just loading it into a dataframe. but with a csv/text file we need to save it somewhere in airflow. probably can be done, but it is easier to deal with items in memory rather that saving it and then loading it. although airflow is already saving files for cumulus and having the file to look at and debug might be useful. Right now I am just talking myself in circles. First step @perrymanmd do you think it would be possible to decouple the parser into a single module? for now the module would just create the out file. We can add the dataframe part later if we want to. Maybe you could then wrap your current process into calling that module and integrate the DSS loader. I could look into this as well next month. Or Brandon if Mike doesn't have time maybe you could take a crack at it. We can schedule a meeting to talk about it as well.

The parser can already be executed without a loader and can output to either stdout or to a specified file in either shefit -1 or shefit -2 format (shefit -1 is better and what the loaders use).

The loaders cannot (yet) be executed standalone but it wouldn't be difficult to:

1. Create a wrapper script that (like `shefParser`) takes loader arguments for instantiation and initialization and then feeds the specified loader from stdin or a specified file (in shefit -1 format).
2. Add a `main()` routine to any of the loaders (e.g., cda_loader) that performed this same logic specifically for that loader.

File: shef_loader/cda_loader.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+        super().__init__(logger, output_object, append)
+        self._cda_url = "https://cwms-data-test.cwbi.us/cwms-data/"
+        self._cwms = CWMSpy.CWMS()
+        self._payloads: list[TimeseriesPayload] = []
+        self._time_series_error_count: int = 0
+        self._transforms: dict[str, ShefTransform] = {}
+        self._value_error_count: int = 0
+        self._write_tasks: list[Coroutine] = []
+
+    def set_options(self, options_str: str | None) -> None:
```
Comment (perrymanmd) [inline shef_loader/cda_loader.py:83]: Change to Union[str, None]

File: shef_loader/cda_loader.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+    ) -> None:
+        """
+        Update error count and output message for failed CDA POST request
+        """
+        self._value_error_count += value_count
+        self._time_series_error_count += 1
+        if self._logger:
+            self._logger.error(f"HTTP {status}: {tsid} - {content.decode()}")
+
+    def find_matching_payload_index(self, payload: TimeseriesPayload) -> int | None:
```
Comment (perrymanmd) [inline shef_loader/cda_loader.py:259]: Change to Union[int, None]

**Review by perrymanmd — CHANGES_REQUESTED** [review]

**Comment by Enovotny** [comment]
ok as I think about this more and more I am coming around to a combined approach. I think having this all as a single python module is the way to go so we don't have to host it in airflow but can just bring the repo in as a package.  So that would mean that we have the following function calls.

1) df = shef _to_df(abc.shef) This would parse the shef file into a dataframe
2) shef_to_file(abc.shef, format =1, outfile = abc.OUT, .....)  This could call shef_to_df and then just save as a file.
3) shef_to_dss(abc.shef)  this would call shef_to_df() and then convert(load) to a dss file
4) shef_to_cwms(abc.shef, apiRoot, apiKey, abc.crit) this would call shef_to_df and the save(load) to the database using CDA.  
5) shef_to...... 

The shef_to_df would be the base parsers script.  the other would call the function and convert to the desired format.  So the shef_to_dss would pass a df to the dss loader to make a dss file.  shef_to_cwms would pass a df to the CDA loader and save into CWMS database.  

The purpose of still decoupling the parsers from the loaders would be again for setting up automated tests as well as ease of interacting with a dataframe and simplifying the loading code.

**Comment by Enovotny** [comment]
we don't have to use dataframes in the above scenario, but some sort of middle product.  My suggesting is dataframes based on my experience, but am open to other suggestions for ease of transition.

**Comment by jbkolze** [comment]
I think we're pretty quickly approaching the point where this will be easier to hash out on a call, but nonetheless...
Eric, I'm gathering from some of your notes that we don't have any easy access to S3 or some other storage mechanism attached to Airflow.  I'm by no means an Airflow expert, but my understanding is that in an ideal world you would:
1. Keep tasks as small as possible/feasible (to make the UI more useful for pinpointing issues)
2. Not pass data between tasks (i.e. pass a filename rather than the file/data itself)
3. Use Airflow purely for orchestration rather than execution

Cumulus is mostly built under this paradigm, e.g. Airflow makes an API call to Cumulus to run grid transform routines rather than running them within Airflow itself, files are stored in S3 between tasks, etc.  As Airflow begins to be used nationally, do we have any concerns with resource utilization within the Airflow environment itself if using it as both an orchestration and execution engine?  If a whole lot of SHEF products are being loaded into CDA using dataframes (or not, honestly) on the hour, I could foresee memory getting eaten up fairly quickly.  And my understanding is that if the scheduler gets low on resources things can get a little wonky...

I am also talking myself in circles here, but I think the point I'm trying to arrive at is that if we're aiming for "best practices" in the long run then a file-based format would probably make more sense than something memory-based (e.g. dataframes).  And, along the same lines, it seems like some sort of file/object storage (S3) should be available to Airflow.  But, it's very possible that we'll never get to the scale that any of this matters, in which case this is all moot.  Or maybe it's just easier to cross that bridge when we get there.

**Comment by jbkolze** [comment]
Also, forgot to mention -- do we foresee many other file formats being parsed / ingested?  I feel like we've switched teams here, Eric, but if there are more than a few other file formats that we could potentially be parsing it seems to me that it would make sense to decouple the loaders from the parsers.  That way if we need to ingest a new file format we only need to build a parser and it will just work with the existing loaders -- and vice versa for a loader.

That being said, I don't really know of any other "standard" formats such as SHEF that we'd be handling off of the top of my head.

File: shef_loader/cda_loader.py (hunk 1 above)
Comment (jbkolze) [reply shef_loader/cda_loader.py]: Fixed in [8df6c69](https://github.com/HydrologicEngineeringCenter/SHEF_processing/pull/9/commits/8df6c697d07a19b6ab781208b1093360c09d13bb)

File: shef_loader/cda_loader.py (hunk 2 above)
Comment (jbkolze) [reply shef_loader/cda_loader.py]: Fixed in [8df6c69](https://github.com/HydrologicEngineeringCenter/SHEF_processing/pull/9/commits/8df6c697d07a19b6ab781208b1093360c09d13bb)

**Review by perrymanmd — APPROVED** [review]

**Pull request merged**